File: app/src/main/webapp/js/classes/function-catalog.js
class FunctionCatalog {
    static fromJson(json) {
        const functionCatalog = new FunctionCatalog();

        const author = new Author();
        author.setId(json.authorId);
        author.setName(json.authorName);

        const company = new Company();
        company.setId(json.companyId);
        company.setName(json.companyName);

        functionCatalog.setId(json.id);
        functionCatalog.setName(json.name);
        functionCatalog.setReleaseVersion(json.releaseVersion);
        functionCatalog.setBaseVersionId(json.baseVersionId);
        functionCatalog.setPriorVersionId(json.priorVersionId);
        functionCatalog.setIsReleased(json.isReleased);
        functionCatalog.setIsApproved(json.isApproved);
        functionCatalog.setApprovalReviewId(json.approvalReviewId);
        functionCatalog.setAuthor(author);
        functionCatalog.setCompany(company);
        functionCatalog.setCreatorAccountId(json.creatorAccountId);
        functionCatalog.setIsDeleted(json.isDeleted);
        functionCatalog.setDeletedDate(json.deletedDate);

        return functionCatalog;
    }

    // Converts existing function catalog into JSON
    static toJson(functionCatalog) {
        const jsonFunctionCatalog = {
            id:                 functionCatalog.getId(),
            name:               functionCatalog.getName(),
            releaseVersion:     functionCatalog.getReleaseVersion(),
            baseVersionId:      functionCatalog.getBaseVersionId(),
            priorVersionId:     functionCatalog.getPriorVersionId(),
            isReleased:         functionCatalog.isReleased(),
            isApproved:         functionCatalog.isApproved(),
            creatorAccountId:          functionCatalog.getCreatorAccountId()
        };
        const author = (functionCatalog.getAuthor() || new Author());
        const company = (functionCatalog.getCompany() || new Company());
        if (author.getId() > 0) {
            jsonFunctionCatalog.authorId = author.getId();
        }
        if (company.getId() > 0) {
            jsonFunctionCatalog.companyId = company.getId();
        }
        return jsonFunctionCatalog;
    }

    constructor() {
        this._id                = null;
        this._name              = "";
        this._releaseVersion    = "";
        this._versionsJson      = null;
        this._isReleased        = null;
        this._isApproved        = null;
        this._priorVersionId    = null;
        this._baseVersionId     = null;
        this._author            = null;
        this._company           = null;
        this._creatorAccountId  = null;
        this._approvalReviewId  = null;
        this._isDeleted         = false;
        this._deletedDate       = null;

        this._functionBlocks    = [];
    };

    setId(id) {
        this._id = id;
    }

    getId() {
        return this._id;
    }

    setName(name) {
        this._name = name;
    }

    getName() {
        return this._name;
    }

    setReleaseVersion(releaseVersion) {
        this._releaseVersion = releaseVersion;
    }

    getReleaseVersion() {
        return this._releaseVersion;
    }

    setAuthor(author) {
        this._author = author;
    }

    getAuthor() {
        return this._author;
    }

    setCompany(company) {
        this._company = company;
    }

    getCompany() {
        return this._company;
    }

    getFunctionBlocks() {
        return this._functionBlocks;
    }

    setVersionsJson(versionsJson) {
        this._versionsJson = versionsJson;
    }

    getVersionsJson() {
        return this._versionsJson;
    }

    setBaseVersionId(baseVersionId) {
        this._baseVersionId = baseVersionId;
    }

    getBaseVersionId() {
        return this._baseVersionId;
    }

    setPriorVersionId(priorVersionId) {
        this._priorVersionId = priorVersionId;
    }

    getPriorVersionId() {
        return this._priorVersionId;
    }

    setIsReleased(isReleased) {
        this._isReleased = isReleased;
    }

    isReleased() {
        return this._isReleased;
    }

    setIsApproved(isApproved) {
        this._isApproved = isApproved;
    }

    isApproved() {
        return this._isApproved;
    }

    setCreatorAccountId(creatorAccountId) {
        this._creatorAccountId = creatorAccountId;
    }

    getCreatorAccountId() {
        return this._creatorAccountId;
    }

    setApprovalReviewId(approvalReviewId) {
        this._approvalReviewId = approvalReviewId;
    }

    getApprovalReviewId() {
        return this._approvalReviewId;
    }

    setIsDeleted(isDeleted) {
        this._isDeleted = isDeleted;
    }

    isDeleted() {
        return this._isDeleted;
    }

    setDeletedDate(deletedDate) {
        this._deletedDate = deletedDate;
    }

    getDeletedDate() {
        return this._deletedDate;
    }

    getDisplayVersion() {
        if (this._isReleased) {
            return this._releaseVersion;
        }
        return this._releaseVersion + "-" + this._id;
    }
}

File: app/src/main/java/com/softwareverde/tidyduck/most/FunctionCatalog.java
package com.softwareverde.tidyduck.most;

import com.softwareverde.mostadapter.Modification;
import com.softwareverde.util.Util;

import java.util.ArrayList;
import java.util.List;

public class FunctionCatalog {
    private Long _id;
    private String _name;
    private String _release;
    private Author _author;
    private Company _company;
    private boolean _isApproved;
    private boolean _isReleased;
    private Long _baseVersionId;
    private Long _priorVersionId;
    private final List<FunctionBlock> _functionBlocks = new ArrayList<>();
    private final List<Modification> _modifications = new ArrayList<>();
    private final List<ClassDefinition> _classDefinitions = new ArrayList<>();
    private final List<PropertyCommandDefinition> _propertyCommandDefinitions = new ArrayList<>();
    private final List<MethodCommandDefinition> _methodCommandDefinitions = new ArrayList<>();
    private final List<PropertyReportDefinition> _propertyReportDefinitions = new ArrayList<>();
    private final List<MethodReportDefinition> _methodReportDefinitions = new ArrayList<>();
    private final List<TypeDefinition> _typeDefinitions = new ArrayList<>();
    private final List<UnitDefinition> _unitDefinitions = new ArrayList<>();
    private final List<ErrorDefinition> _errorDefinitions = new ArrayList<>();

    public Long getId() {
        return _id;
    }

    public void setId(final long id) {
        _id = id;
    }

    public String getName() {
        return _name;
    }

    public void setName(final String name) {
        _name = name;
    }

    public String getRelease() {
        return _release;
    }

    public void setRelease(final String release) {
        _release = release;
    }

    public Author getAuthor() {
        return _author;
    }

    public void setAuthor(final Author author) {
        _author = author;
    }

    public Company getCompany() {
        return _company;
    }

    public void setCompany(final Company company) {
        _company = company;
    }

    public boolean isApproved() {
        return _isApproved;
    }

    public void setIsApproved(final boolean approved) {
        _isApproved = approved;
    }

    public boolean isReleased() {
        return _isReleased;
    }

    public void setIsReleased(final boolean released) {
        _isReleased = released;
    }

    public Long getBaseVersionId() {
        return _baseVersionId;
    }

    public void setBaseVersionId(final Long baseVersionId) {
        _baseVersionId = baseVersionId;
    }

    public Long getPriorVersionId() {
        return _priorVersionId;
    }

    public void setPriorVersionId(final Long priorVersionId) {
        _priorVersionId = priorVersionId;
    }

    public List<FunctionBlock> getFunctionBlocks() {
        return Util.copyList(_functionBlocks);
    }

    public List<ClassDefinition> getClassDefinitions() {
        return Util.copyList(_classDefinitions);
    }

    public List<PropertyCommandDefinition> getPropertyCommandDefinitions() {
        return Util.copyList(_propertyCommandDefinitions);
    }

    public List<MethodCommandDefinition> getMethodCommandDefinitions() {
        return Util.copyList(_methodCommandDefinitions);
    }

    public List<PropertyReportDefinition> getPropertyReportDefinitions() {
        return Util.copyList(_propertyReportDefinitions);
    }

    public List<MethodReportDefinition> getMethodReportDefinitions() {
        return Util.copyList(_methodReportDefinitions);
    }

    public List<TypeDefinition> getTypeDefinitions() {
        return Util.copyList(_typeDefinitions);
    }

    public List<UnitDefinition> getUnitDefinitions() {
        return Util.copyList(_unitDefinitions);
    }

    public List<ErrorDefinition> getErrorDefinitions() {
        return Util.copyList(_errorDefinitions);
    }

    public void addClassDefinition(final ClassDefinition classDefinition) {
        _classDefinitions.add(classDefinition);
    }

    public void setClassDefinitions(final List<ClassDefinition> classDefinitions) {
        _classDefinitions.clear();
        _classDefinitions.addAll(classDefinitions);
    }

    public void addPropertyCommandDefinition(final PropertyCommandDefinition commandDefinition) {
        _propertyCommandDefinitions.add(commandDefinition);
    }

    public void setPropertyCommandDefinitions(final List<PropertyCommandDefinition> commandDefinitions) {
        _propertyCommandDefinitions.clear();
        _propertyCommandDefinitions.addAll(commandDefinitions);
    }

    public void addMethodCommandDefinition(final MethodCommandDefinition commandDefinition) {
        _methodCommandDefinitions.add(commandDefinition);
    }

    public void setMethodCommandDefinitions(final List<MethodCommandDefinition> commandDefinitions) {
        _methodCommandDefinitions.clear();
        _methodCommandDefinitions.addAll(commandDefinitions);
    }

    public void addPropertyReportDefinition(final PropertyReportDefinition reportDefinition) {
        _propertyReportDefinitions.add(reportDefinition);
    }

    public void setPropertyReportDefinitions(final List<PropertyReportDefinition> reportDefinitions) {
        _propertyReportDefinitions.clear();
        _propertyReportDefinitions.addAll(reportDefinitions);
    }

    public void addMethodReportDefinition(final MethodReportDefinition reportDefinition) {
        _methodReportDefinitions.add(reportDefinition);
    }

    public void setMethodReportDefinitions(final List<MethodReportDefinition> reportDefinitions) {
        _methodReportDefinitions.clear();
        _methodReportDefinitions.addAll(reportDefinitions);
    }

    public void addTypeDefinition(final TypeDefinition typeDefinition) {
        _typeDefinitions.add(typeDefinition);
    }

    public void setTypeDefinitions(final List<TypeDefinition> typeDefinitions) {
        _typeDefinitions.clear();
        _typeDefinitions.addAll(typeDefinitions);
    }

    public void addUnitDefinition(final UnitDefinition unitDefinition) {
        _unitDefinitions.add(unitDefinition);
    }

    public void setUnitDefinitions(final List<UnitDefinition> unitDefinitions) {
        _unitDefinitions.clear();
        _unitDefinitions.addAll(unitDefinitions);
    }

    public void addErrorDefinition(final ErrorDefinition errorDefinition) {
        _errorDefinitions.add(errorDefinition);
    }

    public void setErrorDefinitions(final List<ErrorDefinition> errorDefinitions) {
        _errorDefinitions.clear();
        _errorDefinitions.addAll(errorDefinitions);
    }

    public void addFunctionBlock(final FunctionBlock functionBlock) {
        _functionBlocks.add(functionBlock);
    }

    public void setFunctionBlocks(final List<FunctionBlock> functionBlocks) {
        _functionBlocks.clear();
        _functionBlocks.addAll(functionBlocks);
    }

    public List<Modification> getModifications() {
        return Util.copyList(_modifications);
    }

    public void addModification(final Modification modification) {
        _modifications.add(modification);
    }

    public void setModifications(final List<Modification> modifications) {
        _modifications.clear();
        _modifications.addAll(modifications);
    }
}
